### PR TITLE
Adds DoVoid PayPal express NVP method as the gateway's Cancel action.

### PR DESCRIPTION
**Pull request by iisisrael**
Resubmitting the changes mistakenly requested on the read-only [PaypalExpressCheckoutNvp](https://github.com/Payum/PaypalExpressCheckoutNvp/pull/44) subtree split.

### CODE REVIEW

File: src/Payum/Paypal/ExpressCheckout/Nvp/Action/Api/DoVoidAction.php [hunk 1]
```diff
@@ -0,0 +1,61 @@
...
+namespace Payum\Paypal\ExpressCheckout\Nvp\Action\Api;
+
+use Payum\Core\Bridge\Spl\ArrayObject;
+use Payum\Core\Exception\RequestNotSupportedException;
+use Payum\Core\GatewayAwareInterface;
+use Payum\Core\GatewayInterface;
+use Payum\Paypal\ExpressCheckout\Nvp\Request\Api\DoVoid;
+use Payum\Paypal\ExpressCheckout\Nvp\Request\Api\GetTransactionDetails;
+
+class DoVoidAction extends BaseApiAwareAction implements GatewayAwareInterface
```
Comment (makasim) [inline src/Payum/Paypal/ExpressCheckout/Nvp/Action/Api/DoVoidAction.php]: This is better to do this way.

```php
<?php
class DoVoidAction implements ActionInterface, ApiAwareInterface, GatewayAwareInterface
{
   use ApiAwareTrait;
   use GatewayAwareTrait;
   
   public function __construct()
   {
       $this->apiClass = Api:class;
   }
}

File: src/Payum/Paypal/ExpressCheckout/Nvp/Action/Api/DoVoidAction.php [hunk 2]
```diff
@@ -0,0 +1,61 @@
...
+        $fields['AUTHORIZATIONID'] = $parentTransactionId
+            ? $parentTransactionId
+            : $model['PAYMENTREQUEST_'.$paymentRequestN.'_TRANSACTIONID']
+        ;
+
+        $fields->validateNotEmpty(['AUTHORIZATIONID']);
+
+        $this->api->doVoid((array) $fields);
+
+        $this->gateway->execute(new GetTransactionDetails($model, $paymentRequestN));
```
Comment (makasim) [inline src/Payum/Paypal/ExpressCheckout/Nvp/Action/Api/DoVoidAction.php]: this should not be here, instead the CancelAction must do a Sync sub request

File: src/Payum/Paypal/ExpressCheckout/Nvp/Action/CancelAction.php [hunk 3]
```diff
@@ -0,0 +1,43 @@
+<?php
+namespace Payum\Paypal\ExpressCheckout\Nvp\Action;
+
+use Payum\Core\Bridge\Spl\ArrayObject;
+use Payum\Core\Request\Cancel;
+use Payum\Core\Exception\RequestNotSupportedException;
+use Payum\Paypal\ExpressCheckout\Nvp\Api;
+use Payum\Paypal\ExpressCheckout\Nvp\Request\Api\DoVoid;
+
+class CancelAction extends PurchaseAction
```
Comment (makasim) [inline src/Payum/Paypal/ExpressCheckout/Nvp/Action/CancelAction.php]: why it extends PurchaseAction?

**Comment by iisisrael** [comment]
Maybe so, but I duplicated the [DoCaptureAction](https://github.com/Payum/Payum/blob/master/src/Payum/Paypal/ExpressCheckout/Nvp/Action/Api/DoCaptureAction.php) for consistency.  If there are performance or implementation concerns, wouldn't that be better as a separate issue?

File: src/Payum/Paypal/ExpressCheckout/Nvp/Action/Api/DoVoidAction.php (hunk 1 above)
Comment (iisisrael) [inline src/Payum/Paypal/ExpressCheckout/Nvp/Action/Api/DoVoidAction.php]: Maybe so, but I duplicated the [DoCaptureAction](https://github.com/Payum/Payum/blob/master/src/Payum/Paypal/ExpressCheckout/Nvp/Action/Api/DoCaptureAction.php) for consistency.  If there are performance or implementation concerns, wouldn't that be better as a separate issue?

**Comment by makasim** [comment]
> Maybe so, but I duplicated the DoCaptureAction for consistency. If there are performance or implementation concerns, wouldn't that be better as a separate issue?
yeah I know, I am in the process of migration to the new concept, 
in any case the BaseApiAction classes will be removed.

The trait also provide a better exception message when the support api was not found.

File: src/Payum/Paypal/ExpressCheckout/Nvp/Action/Api/DoVoidAction.php (hunk 1 above)
Comment (makasim) [inline src/Payum/Paypal/ExpressCheckout/Nvp/Action/Api/DoVoidAction.php]: > Maybe so, but I duplicated the DoCaptureAction for consistency. If there are performance or implementation concerns, wouldn't that be better as a separate issue?

Yeah I know, I am in the process of migration to the new concept, 
in any case the BaseApiAction classes will be removed.

The trait also provide a better exception message when the support api was not found.

File: src/Payum/Paypal/ExpressCheckout/Nvp/Action/Api/DoVoidAction.php (hunk 2 above)
Comment (iisisrael) [inline src/Payum/Paypal/ExpressCheckout/Nvp/Action/Api/DoVoidAction.php]: I think I see what you mean - fix pushed.

File: src/Payum/Paypal/ExpressCheckout/Nvp/Action/CancelAction.php (hunk 3 above)
Comment (iisisrael) [inline src/Payum/Paypal/ExpressCheckout/Nvp/Action/CancelAction.php]: Lazy copy of CaptureAction?  Fixed to extend GatewayAwareAction and execute a Sync after executing a DoVoid on all payment requests that are Authroized.

File: src/Payum/Paypal/ExpressCheckout/Nvp/Api.php [hunk 4]
```diff
@@ -529,6 +534,26 @@ public function doCapture(array $fields)
...
+     *
+     * @param array $fields
+     *
+     * @return array
+     */
+    public function doVoid(array $fields)
+    {
+        $fields['METHOD']  = 'DoVoid';
+        if (empty($fields['AUTHORIZATIONID'])) {
+            $fields['AUTHORIZATIONID'] = $fields['TRANSACTIONID'];
```
Comment (makasim) [inline src/Payum/Paypal/ExpressCheckout/Nvp/Api.php]: this is logic must be moved to action, it is not api related stuff

File: src/Payum/Paypal/ExpressCheckout/Nvp/Action/Api/DoVoidAction.php (hunk 1 above)
Comment (makasim) [inline src/Payum/Paypal/ExpressCheckout/Nvp/Action/Api/DoVoidAction.php]: @iisisrael Could you please use traits approach

Comment (iisisrael) [inline src/Payum/Paypal/ExpressCheckout/Nvp/Action/Api/DoVoidAction.php]: Done, thanks for the code snippet.

File: src/Payum/Paypal/ExpressCheckout/Nvp/Api.php (hunk 4 above)
Comment (iisisrael) [inline src/Payum/Paypal/ExpressCheckout/Nvp/Api.php]: Done.  I also noticed the action was missing tests, so added those.  I think I covered all the cases.

**Pull request merged**